File: client/mia_react_native/src/services/realtimeCallService.ts
import LiveAudioStream from 'react-native-live-audio-stream';
import { PCM_BYTES_PER_SECOND, PCM_SAMPLE_RATE, PcmPlayer } from './pcmPlayer';

export type CallConnectionState = 'connecting' | 'ready' | 'error' | 'ended';

type Listener<T> = (value: T) => void;

const WS_PING_MS = 20_000;
const PLAYBACK_DRAIN_PADDING_MS = 900;
const EMERGENCY_FALLBACK_MS = 12_000;

function base64ToBytes(b64: string): Uint8Array {
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}

function bytesToBase64(bytes: Uint8Array): string {
  let binary = '';
  for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i]);
  return btoa(binary);
}

function rmsLevel(bytes: Uint8Array): number {
  if (bytes.length < 2) return 8;
  let sum = 0;
  for (let i = 0; i < bytes.length - 1; i += 2) {
    const sample = bytes[i] | (bytes[i + 1] << 8);
    const signed = sample > 32767 ? sample - 65536 : sample;
    sum += signed * signed;
  }
  const rms = Math.sqrt(sum / (bytes.length / 2));
  return Math.min(56, Math.max(6, Math.round(rms)));
}

/** xAI Realtime voice call — PCM 24 kHz mono over WebSocket. Half-duplex on mobile. */
export class RealtimeCallService {
  private ws: WebSocket | null = null;
  private player = new PcmPlayer();

  private connected = false;
  private sessionReady = false;
  private muted = false;
  private speakerOn = false;
  private assistantSpeaking = false;
  private micRunning = false;
  private wantMicRunning = false;
  private fatalError = false;
  private assistantAudioBytes = 0;
  private assistantStartedAt: number | null = null;
  private assistantFinishTimer: ReturnType<typeof setTimeout> | null = null;
  private sessionReadyResolve: (() => void) | null = null;
  private sessionReadyPromise: Promise<void> | null = null;
  private micDataHandler: ((data: string) => void) | null = null;

  private transcriptListeners = new Set<Listener<string>>();
  private levelListeners = new Set<Listener<number>>();
  private connectionListeners = new Set<Listener<CallConnectionState>>();

  onTranscript(listener: Listener<string>): () => void {
    this.transcriptListeners.add(listener);
    return () => this.transcriptListeners.delete(listener);
  }

  onLevel(listener: Listener<number>): () => void {
    this.levelListeners.add(listener);
    return () => this.levelListeners.delete(listener);
  }

  onConnection(listener: Listener<CallConnectionState>): () => void {
    this.connectionListeners.add(listener);
    return () => this.connectionListeners.delete(listener);
  }

  private emitTranscript(t: string): void {
    this.transcriptListeners.forEach((l) => l(t));
  }

  private emitLevel(n: number): void {
    this.levelListeners.forEach((l) => l(n));
  }

  private emitConnection(s: CallConnectionState): void {
    this.connectionListeners.forEach((l) => l(s));
  }

  get isConnected(): boolean {
    return this.connected && this.sessionReady;
  }

  async connect(options: {
    wsUrl: string;
    token: string;
    sessionConfig: Record<string, unknown>;
    sessionPreconfigured?: boolean;
  }): Promise<void> {
    this.fatalError = false;
    this.sessionReady = false;
    this.assistantSpeaking = false;
    this.assistantAudioBytes = 0;
    this.assistantStartedAt = null;
    this.sessionReadyPromise = new Promise((resolve) => {
      this.sessionReadyResolve = resolve;
    });
    this.emitConnection('connecting');

    await this.player.setup();

type HeaderWebSocket = new (
  url: string,
  protocols?: string | string[],
  options?: { headers?: Record<string, string> },
) => WebSocket;

    const WS = WebSocket as HeaderWebSocket;
    this.ws = new WS(options.wsUrl, undefined, {
      headers: { Authorization: `Bearer ${options.token}` },
    });

    this.connected = true;

    this.ws.onmessage = (ev) => this.onEvent(ev.data);
    this.ws.onerror = () => this.emitError('connection lost');
    this.ws.onclose = () => {
      if (this.connected && !this.fatalError) this.emitError('call disconnected');
      this.connected = false;
    };

    if (!options.sessionPreconfigured) {
      await new Promise((r) => setTimeout(r, 300));
      const session = { ...options.sessionConfig };
      delete session.model;
      this.send({ type: 'session.update', session });
    }

    await Promise.race([
      this.sessionReadyPromise,
      new Promise<void>((_, reject) =>
        setTimeout(() => {
          if (!this.sessionReady && !this.fatalError) this.emitError('voice session timed out');
          reject(new Error('timeout'));
        }, 12_000),
      ),
    ]).catch(() => {});

    if (this.fatalError || !this.sessionReady) {
      throw new Error(this.fatalError ? 'voice call failed' : 'voice session not ready');
    }

    this.wantMicRunning = !this.muted;
    await this.applyMicState();
    this.emitConnection('ready');
  }

  async setMuted(muted: boolean): Promise<void> {
    this.muted = muted;
    this.wantMicRunning = this.connected && this.sessionReady && !this.muted && !this.assistantSpeaking;
    await this.applyMicState();
    if (muted && this.connected) this.send({ type: 'input_audio_buffer.clear' });
  }

  async setSpeaker(on: boolean): Promise<void> {
    this.speakerOn = on;
  }

  private send(payload: Record<string, unknown>): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;
    try {
      this.ws.send(JSON.stringify(payload));
    } catch {
      /* ignore */
    }
  }

  private markSessionReady(): void {
    if (this.sessionReady) return;
    this.sessionReady = true;
    this.sessionReadyResolve?.();
  }

  private beginAssistantTurn(): void {
    if (this.assistantSpeaking) return;
    this.assistantSpeaking = true;
    this.assistantAudioBytes = 0;
    this.assistantStartedAt = Date.now();
    if (this.assistantFinishTimer) clearTimeout(this.assistantFinishTimer);
    this.wantMicRunning = false;
    void this.applyMicState();
    this.send({ type: 'input_audio_buffer.clear' });
    this.scheduleAssistantFinish(EMERGENCY_FALLBACK_MS);
  }

  private scheduleAssistantFinish(delayMs: number): void {
    if (this.assistantFinishTimer) clearTimeout(this.assistantFinishTimer);
    this.assistantFinishTimer = setTimeout(() => this.finishAssistantTurn(), delayMs);
  }

  private finishAssistantTurn(): void {
    if (this.assistantFinishTimer) clearTimeout(this.assistantFinishTimer);
    if (!this.assistantSpeaking) return;
    this.assistantSpeaking = false;
    this.assistantAudioBytes = 0;
    this.assistantStartedAt = null;
    this.send({ type: 'input_audio_buffer.clear' });
    this.wantMicRunning = this.connected && this.sessionReady && !this.muted;
    void this.applyMicState();
    this.emitTranscript('…say something — i\'m listening');
  }

  private rescheduleFinishFromPlayback(): void {
    if (!this.assistantSpeaking) return;
    const startedAt = this.assistantStartedAt ?? Date.now();
    const elapsedMs = Date.now() - startedAt;
    const audioMs = Math.ceil((this.assistantAudioBytes / PCM_BYTES_PER_SECOND) * 1000);
    const remainingMs = Math.max(0, audioMs - elapsedMs);
    this.scheduleAssistantFinish(remainingMs + PLAYBACK_DRAIN_PADDING_MS);
  }

  private emitError(message: string): void {
    this.fatalError = true;
    this.emitConnection('error');
    this.emitTranscript(message);
    this.sessionReadyResolve?.();
  }

  private async applyMicState(): Promise<void> {
    while (this.micRunning !== this.wantMicRunning) {
      if (this.wantMicRunning) await this.startMicNow();
      else await this.stopMicNow();
    }
  }

  private async startMicNow(): Promise<void> {
    if (this.micRunning) return;

    LiveAudioStream.init({
      sampleRate: PCM_SAMPLE_RATE,
      channels: 1,
      bitsPerSample: 16,
      audioSource: 6,
      bufferSize: 4096,
      wavFile: '',
    });

    this.micDataHandler = (data: string) => {
      if (!this.connected || !this.sessionReady || this.muted || this.assistantSpeaking) return;
      const chunk = base64ToBytes(data);
      this.emitLevel(rmsLevel(chunk));
      this.send({ type: 'input_audio_buffer.append', audio: data });
    };

    LiveAudioStream.on('data', this.micDataHandler);
    LiveAudioStream.start();
    this.micRunning = true;
  }

  private async stopMicNow(): Promise<void> {
    if (!this.micRunning && !this.micDataHandler) return;
    try {
      LiveAudioStream.stop();
    } catch {
      /* ignore */
    }
    this.micDataHandler = null;
    this.micRunning = false;
  }

  private onEvent(raw: unknown): void {
    if (typeof raw !== 'string') return;
    let event: Record<string, unknown>;
    try {
      event = JSON.parse(raw) as Record<string, unknown>;
    } catch {
      return;
    }

    const type = event.type as string | undefined;

    switch (type) {
      case 'session.created':
      case 'session.updated':
      case 'conversation.created':
        this.markSessionReady();
        break;

      case 'response.output_audio.delta':
      case 'response.audio.delta': {
        const b64 = event.delta as string | undefined;
        if (b64) {
          this.beginAssistantTurn();
          const pcm = base64ToBytes(b64);
          this.assistantAudioBytes += pcm.length;
          this.player.feed(pcm);
          this.rescheduleFinishFromPlayback();
        }
        break;
      }

      case 'response.output_audio_transcript.delta': {
        const delta = (event.delta as string) ?? '';
        if (delta) this.emitTranscript(delta);
        break;
      }
      case 'response.output_audio_transcript.done': {
        const done = (event.transcript as string) ?? '';
        if (done) this.emitTranscript(done);
        break;
      }

      case 'input_audio_buffer.speech_started':
        this.emitLevel(40);
        this.emitTranscript('…i heard you');
        break;
      case 'input_audio_buffer.speech_stopped':
        this.emitTranscript('…one sec, mia is thinking');
        break;

      case 'response.done':
      case 'response.output_audio.done':
        this.rescheduleFinishFromPlayback();
        break;

      case 'response.cancelled':
        this.finishAssistantTurn();
        break;

      case 'error': {
        const err = event.error;
        const msg =
          err && typeof err === 'object' && 'message' in err
            ? String((err as { message?: string }).message ?? err)
            : String(event);
        this.emitError(msg);
        break;
      }
    }
  }

  async hangUp(): Promise<void> {
    this.connected = false;
    this.sessionReady = false;
    this.assistantSpeaking = false;
    if (this.assistantFinishTimer) clearTimeout(this.assistantFinishTimer);
    this.wantMicRunning = false;
    await this.applyMicState();
    try {
      this.ws?.close();
    } catch {
      /* ignore */
    }
    this.ws = null;
    await this.player.release();
    this.emitConnection('ended');
  }
}

File: client/mia_companion/lib/services/realtime_call_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io' show Platform;
import 'dart:math' as math;
import 'dart:typed_data' show ByteData, Uint8List;

import 'package:audio_session/audio_session.dart';
import 'package:flutter_pcm_sound/flutter_pcm_sound.dart';
import 'package:record/record.dart' hide IosAudioCategory;
import 'package:web_socket_channel/io.dart';

import 'debug_log.dart';

enum CallConnectionState { connecting, ready, error, ended }

/// xAI Realtime voice call — PCM 24 kHz mono over WebSocket.
///
/// This is deliberately half-duplex on mobile: while Mia is speaking, the
/// recorder is stopped. That prevents loudspeaker audio from being sent back
/// to xAI as user speech, which was causing echo loops and stalled turns.
class RealtimeCallService {
  static const Duration _wsPingInterval = Duration(seconds: 20);
  static const int _sampleRate = 24000;
  static const int _pcmBytesPerSecond = _sampleRate * 2; // mono PCM16
  /// Extra padding (ms) added to the computed audio duration so the speaker
  /// fully drains before the mic resumes. Kept generous so the mic does not
  /// reopen during short pauses in Zara's speech (which would let her audio
  /// echo back into the mic and chop her reply into pieces).
  static const int _playbackDrainPaddingMs = 1400;
  /// Used only if no audio delta ever arrives for a "response" — should not
  /// normally fire because every delta reschedules the finish timer based on
  /// actual queued audio duration.
  static const Duration _emergencyFallback = Duration(seconds: 12);

  IOWebSocketChannel? _channel;
  StreamSubscription<dynamic>? _wsSub;
  StreamSubscription<Uint8List>? _micSub;
  final AudioRecorder _recorder = AudioRecorder();
  AudioSession? _audioSession;

  final StreamController<String> transcriptController =
      StreamController.broadcast();
  final StreamController<int> levelController = StreamController.broadcast();
  final StreamController<CallConnectionState> connectionController =
      StreamController.broadcast();

  bool _connected = false;
  bool _sessionReady = false;
  bool _muted = false;
  bool _speakerOn = false;
  bool _assistantSpeaking = false;
  bool _micRunning = false;
  bool _wantMicRunning = false;
  bool _fatalError = false;
  int _micChunkCount = 0;
  int _micCycleId = 0;
  int _assistantAudioBytes = 0;
  DateTime? _assistantStartedAt;
  Timer? _assistantFinishTimer;
  Future<void> _micQueue = Future<void>.value();
  Completer<void>? _sessionReadyCompleter;

  bool get isConnected => _connected && _sessionReady;

  Future<void> connect({
    required String wsUrl,
    required String token,
    required Map<String, dynamic> sessionConfig,
    bool sessionPreconfigured = false,
  }) async {
    _fatalError = false;
    _sessionReady = false;
    _assistantSpeaking = false;
    _assistantAudioBytes = 0;
    _assistantStartedAt = null;
    _sessionReadyCompleter = Completer<void>();
    connectionController.add(CallConnectionState.connecting);

    await _configureAudioSession();
    await _initPlayback();

    _channel = IOWebSocketChannel.connect(
      Uri.parse(wsUrl),
      headers: {'Authorization': 'Bearer $token'},
      pingInterval: _wsPingInterval,
    );
    _connected = true;

    _wsSub = _channel!.stream.listen(
      _onEvent,
      onError: (e) {
        // #region agent log
        DebugLog.send(
          location: 'realtime_call_service.dart:wsOnError',
          message: 'WebSocket onError fired',
          hypothesisId: 'H-A,H-E',
          data: {'error': e.toString()},
        );
        // #endregion
        _emitError('connection lost: $e');
      },
      onDone: () {
        // #region agent log
        DebugLog.send(
          location: 'realtime_call_service.dart:wsOnDone',
          message: 'WebSocket onDone fired',
          hypothesisId: 'H-A,H-E',
          data: {
            'wasConnected': _connected,
            'fatalError': _fatalError,
            'closeCode': _channel?.closeCode,
            'closeReason': _channel?.closeReason,
          },
        );
        // #endregion
        if (_connected && !_fatalError) {
          _emitError('call disconnected');
        }
        _connected = false;
      },
    );

    if (!sessionPreconfigured) {
      await Future<void>.delayed(const Duration(milliseconds: 300));
      final session = Map<String, dynamic>.from(sessionConfig)
        ..remove('model');
      _send({'type': 'session.update', 'session': session});
    }

    await _sessionReadyCompleter!.future.timeout(
      const Duration(seconds: 12),
      onTimeout: () {
        if (!_sessionReady && !_fatalError) {
          _emitError('voice session timed out');
        }
      },
    );

    if (_fatalError || !_sessionReady) {
      throw Exception(
        _fatalError ? 'voice call failed' : 'voice session not ready',
      );
    }

    _wantMicRunning = !_muted;
    await _applyMicState();
    connectionController.add(CallConnectionState.ready);
  }

  Future<void> _initPlayback() async {
    await FlutterPcmSound.setLogLevel(LogLevel.error);
    await FlutterPcmSound.setup(
      sampleRate: _sampleRate,
      channelCount: 1,
      iosAudioCategory: IosAudioCategory.playAndRecord,
    );
    await FlutterPcmSound.setFeedThreshold(2400);
    FlutterPcmSound.setFeedCallback((_) {});
    FlutterPcmSound.start();
  }

  Future<void> _configureAudioSession() async {
    _audioSession = await AudioSession.instance;
    await _audioSession!.configure(
      AudioSessionConfiguration(
        avAudioSessionCategory: AVAudioSessionCategory.playAndRecord,
        avAudioSessionCategoryOptions: _speakerOn
            ? AVAudioSessionCategoryOptions.defaultToSpeaker |
                AVAudioSessionCategoryOptions.allowBluetooth
            : AVAudioSessionCategoryOptions.allowBluetooth,
        avAudioSessionMode: AVAudioSessionMode.voiceChat,
        androidAudioAttributes: const AndroidAudioAttributes(
          contentType: AndroidAudioContentType.speech,
          usage: AndroidAudioUsage.voiceCommunication,
        ),
        androidAudioFocusGainType: AndroidAudioFocusGainType.gain,
        androidWillPauseWhenDucked: false,
      ),
    );
    await _audioSession!.setActive(true);
    if (Platform.isAndroid) {
      await AndroidAudioManager().setSpeakerphoneOn(_speakerOn);
    }
  }

  Future<void> setMuted(bool muted) async {
    _muted = muted;
    _wantMicRunning =
        _connected && _sessionReady && !_muted && !_assistantSpeaking;
    _queueMicStateSync();
    if (muted && _connected) {
      _send({'type': 'input_audio_buffer.clear'});
    }
  }

  Future<void> setSpeaker(bool on) async {
    _speakerOn = on;
    await _configureAudioSession();
  }

  void _send(Map<String, dynamic> payload) {
    final ch = _channel;
    if (ch == null) {
      // #region agent log
      DebugLog.send(
        location: 'realtime_call_service.dart:_send',
        message: 'Tried to send but channel is null',
        hypothesisId: 'H-A,H-E',
        data: {'type': payload['type']},
      );
      // #endregion
      return;
    }
    try {
      ch.sink.add(jsonEncode(payload));
    } catch (e) {
      // #region agent log
      DebugLog.send(
        location: 'realtime_call_service.dart:_send',
        message: 'sink.add THREW',
        hypothesisId: 'H-A,H-E',
        data: {'type': payload['type'], 'error': e.toString()},
      );
      // #endregion
    }
  }

  void _markSessionReady() {
    if (_sessionReady) return;
    _sessionReady = true;
    _sessionReadyCompleter?.complete();
  }

  void _beginAssistantTurn() {
    if (_assistantSpeaking) return;
    _assistantSpeaking = true;
    _assistantAudioBytes = 0;
    _assistantStartedAt = DateTime.now();
    _assistantFinishTimer?.cancel();
    _wantMicRunning = false;
    _queueMicStateSync();
    _send({'type': 'input_audio_buffer.clear'});
    // Emergency fallback so the turn always ends even if something goes wrong
    // with delta accounting. Each subsequent delta will replace this timer
    // with the duration-based schedule.
    _scheduleAssistantFinish(_emergencyFallback);
    // #region agent log
    DebugLog.send(
      location: 'realtime_call_service.dart:_beginAssistantTurn',
      message: 'Assistant turn started — mic will be stopped',
      hypothesisId: 'H-B,H-C,H-F',
      data: {'micRunning': _micRunning, 'wantMicRunning': _wantMicRunning},
    );
    // #endregion
  }

  void _scheduleAssistantFinish(Duration delay) {
    _assistantFinishTimer?.cancel();
    _assistantFinishTimer = Timer(delay, _finishAssistantTurn);
  }

  void _finishAssistantTurn() {
    _assistantFinishTimer?.cancel();
    if (!_assistantSpeaking) return;
    _assistantSpeaking = false;
    _assistantAudioBytes = 0;
    _assistantStartedAt = null;
    _send({'type': 'input_audio_buffer.clear'});
    _wantMicRunning = _connected && _sessionReady && !_muted;
    _queueMicStateSync();
    transcriptController.add('…say something — i\'m listening');
    // #region agent log
    DebugLog.send(
      location: 'realtime_call_service.dart:_finishAssistantTurn',
      message: 'Assistant turn finished — mic will restart',
      hypothesisId: 'H-B,H-C',
      data: {
        'connected': _connected,
        'sessionReady': _sessionReady,
        'muted': _muted,
        'wantMicRunning': _wantMicRunning,
        'micRunning': _micRunning,
      },
    );
    // #endregion
  }

  /// Compute when the speaker buffer will be empty, then schedule the finish
  /// for that moment + a small drain pad. Called on every audio delta — the
  /// most recently scheduled timer always wins.
  void _rescheduleFinishFromPlayback() {
    if (!_assistantSpeaking) return;
    final startedAt = _assistantStartedAt;
    final elapsedMs = startedAt == null
        ? 0
        : DateTime.now().difference(startedAt).inMilliseconds;
    final audioMs = (_assistantAudioBytes / _pcmBytesPerSecond * 1000).ceil();
    final remainingMs = math.max(0, audioMs - elapsedMs);
    final drainMs = remainingMs + _playbackDrainPaddingMs;
    _scheduleAssistantFinish(Duration(milliseconds: drainMs));
    // #region agent log
    DebugLog.send(
      location: 'realtime_call_service.dart:_rescheduleFinishFromPlayback',
      message: 'Recomputed finish schedule from playback duration',
      hypothesisId: 'H-F',
      data: {
        'audioMs': audioMs,
        'elapsedMs': elapsedMs,
        'remainingMs': remainingMs,
        'drainMs': drainMs,
      },
    );
    // #endregion
  }

  void _emitError(String message) {
    // #region agent log
    DebugLog.send(
      location: 'realtime_call_service.dart:_emitError',
      message: 'Fatal error path',
      hypothesisId: 'H-G',
      data: {'msg': message},
    );
    // #endregion
    _fatalError = true;
    connectionController.add(CallConnectionState.error);
    transcriptController.add(message);
    if (_sessionReadyCompleter != null &&
        !_sessionReadyCompleter!.isCompleted) {
      _sessionReadyCompleter!.complete();
    }
  }

  void _queueMicStateSync() {
    _micQueue = _micQueue.catchError((_) {}).then((_) => _applyMicState());
  }

  Future<void> _applyMicState() async {
    while (_micRunning != _wantMicRunning) {
      if (_wantMicRunning) {
        await _startMicNow();
      } else {
        await _stopMicNow();
      }
    }
  }

  Future<void> _startMicNow() async {
    if (_micRunning) return;
    if (!await _recorder.hasPermission()) {
      throw Exception('Microphone permission denied');
    }
    Stream<Uint8List> stream;
    try {
      stream = await _recorder.startStream(
        RecordConfig(
          encoder: AudioEncoder.pcm16bits,
          sampleRate: _sampleRate,
          numChannels: 1,
          echoCancel: true,
          noiseSuppress: true,
          autoGain: true,
          androidConfig: AndroidRecordConfig(
            audioSource: AndroidAudioSource.voiceCommunication,
            audioManagerMode: AudioManagerMode.modeInCommunication,
            speakerphone: _speakerOn,
          ),
        ),
      );
      // #region agent log
      DebugLog.send(
        location: 'realtime_call_service.dart:_startMicNow',
        message: 'Recorder startStream succeeded',
        hypothesisId: 'H-B',
        data: {},
      );
      // #endregion
    } catch (e) {
      // #region agent log
      DebugLog.send(
        location: 'realtime_call_service.dart:_startMicNow',
        message: 'Recorder startStream FAILED',
        hypothesisId: 'H-B',
        data: {'error': e.toString()},
      );
      // #endregion
      rethrow;
    }
    _micRunning = true;
    _micChunkCount = 0;
    final cycleId = ++_micCycleId;
    _micSub = stream.listen(
      (chunk) {
        if (!_connected || !_sessionReady || _muted || _assistantSpeaking) {
          return;
        }
        levelController.add(_rmsLevel(chunk));
        _micChunkCount++;
        // #region agent log
        if (_micChunkCount == 1 || _micChunkCount % 10 == 0) {
          DebugLog.send(
            location: 'realtime_call_service.dart:micChunk',
            message: 'Mic chunk sent to xAI',
            hypothesisId: 'H-D,H-E,H-G',
            data: {
              'cycleId': cycleId,
              'chunkNum': _micChunkCount,
              'chunkBytes': chunk.length,
              'rms': _rmsLevel(chunk),
            },
          );
        }
        // #endregion
        _send({
          'type': 'input_audio_buffer.append',
          'audio': base64Encode(chunk),
        });
      },
      onError: (e) {
        // #region agent log
        DebugLog.send(
          location: 'realtime_call_service.dart:micStreamOnError',
          message: 'Mic stream emitted error',
          hypothesisId: 'H-G',
          data: {'cycleId': cycleId, 'error': e.toString()},
        );
        // #endregion
        _emitError('mic error: $e');
      },
      onDone: () {
        // #region agent log
        DebugLog.send(
          location: 'realtime_call_service.dart:micStreamOnDone',
          message: 'Mic stream closed unexpectedly',
          hypothesisId: 'H-G',
          data: {
            'cycleId': cycleId,
            'chunksDelivered': _micChunkCount,
            'micRunning': _micRunning,
          },
        );
        // #endregion
      },
    );
  }

  Future<void> _stopMicNow() async {
    if (!_micRunning && _micSub == null) return;
    await _micSub?.cancel();
    _micSub = null;
    Object? stopError;
    try {
      await _recorder.stop();
    } catch (e) {
      stopError = e;
    }
    _micRunning = false;
    // #region agent log
    DebugLog.send(
      location: 'realtime_call_service.dart:_stopMicNow',
      message: 'Recorder stopped',
      hypothesisId: 'H-B',
      data: {'stopError': stopError?.toString()},
    );
    // #endregion
  }

  int _rmsLevel(Uint8List bytes) {
    if (bytes.length < 2) return 8;
    var sum = 0.0;
    for (var i = 0; i < bytes.length - 1; i += 2) {
      final sample = bytes[i] | (bytes[i + 1] << 8);
      final signed = sample > 32767 ? sample - 65536 : sample;
      sum += signed * signed;
    }
    final rms = math.sqrt(sum / (bytes.length / 2));
    return rms.clamp(6, 56).toInt();
  }

  void _onEvent(dynamic raw) {
    if (raw is! String) return;
    Map<String, dynamic> event;
    try {
      event = jsonDecode(raw) as Map<String, dynamic>;
    } catch (_) {
      return;
    }

    final type = event['type'] as String?;
    // #region agent log
    if (type != 'response.output_audio.delta' &&
        type != 'response.audio.delta' &&
        type != 'response.output_audio_transcript.delta') {
      DebugLog.send(
        location: 'realtime_call_service.dart:_onEvent',
        message: 'WS event received',
        hypothesisId: 'H-A,H-C,H-D',
        data: {'type': type ?? 'unknown'},
      );
    }
    // #endregion

    switch (type) {
      case 'session.created':
      case 'session.updated':
      case 'conversation.created':
        _markSessionReady();
        break;

      case 'response.output_audio.delta':
      case 'response.audio.delta':
        final b64 = event['delta'] as String?;
        if (b64 != null && b64.isNotEmpty) {
          _beginAssistantTurn();
          // Decode synchronously so we can update the audio-duration counter
          // BEFORE scheduling the finish timer. Feeding the speaker is async.
          final pcm = base64Decode(b64);
          _assistantAudioBytes += pcm.length;
          unawaited(_feedPcm(pcm));
          _rescheduleFinishFromPlayback();
        }
        break;

      case 'response.output_audio_transcript.delta':
        final delta = event['delta'] as String? ?? '';
        if (delta.isNotEmpty) transcriptController.add(delta);
        break;
      case 'response.output_audio_transcript.done':
        final done = event['transcript'] as String? ?? '';
        if (done.isNotEmpty) transcriptController.add(done);
        break;

      case 'input_audio_buffer.speech_started':
        levelController.add(40);
        transcriptController.add('…i heard you');
        break;
      case 'input_audio_buffer.speech_stopped':
        transcriptController.add('…one sec, zara is thinking');
        break;

      case 'response.done':
      case 'response.output_audio.done':
        // If server actually sends these, recompute one last time so any
        // tail audio still in the speaker queue plays out fully.
        _rescheduleFinishFromPlayback();
        break;

      case 'response.cancelled':
        _finishAssistantTurn();
        break;

      case 'error':
        final err = event['error'];
        final msg = err is Map
            ? (err['message'] as String? ?? err.toString())
            : event.toString();
        _emitError(msg);
        break;
    }
  }

  Future<void> _feedPcm(Uint8List pcm) async {
    try {
      if (pcm.isEmpty) return;
      await FlutterPcmSound.feed(
        PcmArrayInt16(bytes: ByteData.sublistView(pcm)),
      );
    } catch (_) {
      // Ignore individual chunk feed failures.
    }
  }

  Future<void> hangUp() async {
    _connected = false;
    _sessionReady = false;
    _assistantSpeaking = false;
    _assistantFinishTimer?.cancel();
    _wantMicRunning = false;
    await _applyMicState();
    await _wsSub?.cancel();
    _wsSub = null;
    await _channel?.sink.close();
    _channel = null;
    try {
      await FlutterPcmSound.release();
    } catch (_) {}
    try {
      await _audioSession?.setActive(false);
    } catch (_) {}
    connectionController.add(CallConnectionState.ended);
  }

  void dispose() {
    transcriptController.close();
    levelController.close();
    connectionController.close();
  }
}
